### PR TITLE
#0: Fixes for TensorLayout fromLegacyPaddedShape

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
Tensor creation failed for some shape due to incorrect derivation of alignment from legacy padded shape

### What's changed
Use default alignment if logical shape and padded shape are the same
Better handling for 0-volume tensors in TensorLayout
Added tests for creating tensors with different shapes

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/11621126032)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp [hunk 1]
```diff
@@ -71,8 +76,38 @@ INSTANTIATE_TEST_SUITE_P(
...
+    EmptyTensorTestWithShape,
+    EmptyTensorTest,
+    ::testing::Values(
+        EmptyTensorParams{.shape=ttnn::Shape({1, 1, 32, 32})},
+        EmptyTensorParams{.shape=ttnn::Shape({2, 1, 32, 32})},
+        EmptyTensorParams{.shape=ttnn::Shape({64, 1, 256, 1})},
+        EmptyTensorParams{.shape=ttnn::Shape({1, 1, 21120, 16})},
+        EmptyTensorParams{.shape=ttnn::Shape({0, 0, 0, 0})},
+        EmptyTensorParams{.shape=ttnn::Shape({0, 1, 32, 32})},
+        EmptyTensorParams{.shape=ttnn::Shape({0})}
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp:111]: wdyt about rank 0?

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp [hunk 2]
```diff
@@ -79,10 +84,10 @@ std::optional<ShardSpecBuffer> TensorLayout::compute_shard_spec_buffer(const ttn
     const Size physical_size = compute_physical_shape(shape);
     const Size page_shape = compute_page_shape(physical_size);
 
-    TT_FATAL(physical_size.width() % page_shape.width() == 0, "Physical width {} must be multiple of page width {}", physical_size.width(), page_shape.width());
-    TT_FATAL(physical_size.height() % page_shape.height() == 0, "Physical height {} must be multiple of page height {}", physical_size.height(), page_shape.height());
-    const auto width_in_pages = physical_size.width() / page_shape.width();
-    const auto height_in_pages = physical_size.height() / page_shape.height();
+    TT_FATAL(page_shape.width() == 0 ? physical_size.width() == 0 : physical_size.width() % page_shape.width() == 0, "Physical width {} must be multiple of page width {}", physical_size.width(), page_shape.width());
+    TT_FATAL(page_shape.height() == 0 ? physical_size.height() == 0 : physical_size.height() % page_shape.height() == 0, "Physical height {} must be multiple of page height {}", physical_size.height(), page_shape.height());
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:88]: Can you revisit this check? Very hard to understand

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp [hunk 3]
```diff
@@ -79,10 +84,10 @@ std::optional<ShardSpecBuffer> TensorLayout::compute_shard_spec_buffer(const ttn
...
+    const auto width_in_pages = page_shape.width() == 0 ? 0 : physical_size.width() / page_shape.width();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:89]: Would it be better if we made sure that page_shape is never 0?

File: ttnn/cpp/ttnn/tensor/tensor_impl.cpp [hunk 4]
```diff
@@ -755,16 +755,16 @@ Tensor to_device(const Tensor& tensor, Device* target_device, const MemoryConfig
     TT_FATAL(target_device != nullptr, "Need target device in order to move tensor to device!");
     TT_FATAL(tensor.is_allocated(), "Need data to exist in order to move it to device");
 
+    auto legacy_shape = tensor.get_shape();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor_impl.cpp:758]: This actually returns ttnn::Shape, which we call Shape, and not LegacyShape. Lets tweak names here a bit

File: ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp [hunk 5]
```diff
@@ -30,7 +35,7 @@ Alignment legacyPaddedShapeToAlignment(const ttnn::SimpleShape& legacy_padded_sh
         values[i] = legacy_padded_shape[i] * values[i + 1];
     }
 
-    Alignment result(values);
+    Alignment result(std::move(values));
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/layout/tensor_layout.cpp:38]: Thanks!

File: tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp (hunk 1 above)
Comment (sminakov-tt) [reply tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp]: Added tests for 1-volume tensors, including rank 0

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**